cpp/arcticdb/pipeline/write_frame.cpp

```diff
@@ -33,142 +33,195 @@ namespace arcticdb::pipelines {
                 if (opt_error.has_value()) {
-                    opt_error->raise(frame.desc.fields(0).name(), offset_in_frame);
+                    opt_error->raise(frame_->desc.fields(0).name(), offset_in_frame);
                 }
             }
 
-            for (size_t col = 0, end = slice.col_range.diff(); col < end; ++col) {
-                auto abs_col = col + frame.desc.index().field_count();
-                auto &fd = slice.non_index_field(col);
-                auto &tensor = frame.field_tensors[slice.absolute_field_col(col)];
+            for (size_t col = 0, end = slice_.col_range.diff(); col < end; ++col) {
+                auto abs_col = col + frame_->desc.index().field_count();
+                auto& fd = slice_.non_index_field(col);
+                auto& tensor = frame_->field_tensors[slice_.absolute_field_col(col)];
                 auto opt_error = aggregator_set_data(
                     fd.type(),
-                    tensor, agg, abs_col, rows_to_write, offset_in_frame, slice_num_for_column,
-                    regular_slice_size, sparsify_floats);
+                    tensor, agg, abs_col, rows_to_write, offset_in_frame, slice_num_for_column_,
+                    regular_slice_size, sparsify_floats_);
                 if (opt_error.has_value()) {
                     opt_error->raise(fd.name(), offset_in_frame);
                 }
             }
 
-            ++slice_num_for_column;
             agg.end_block_write(rows_to_write);
             agg.commit();
-        }
-    });
+            return output;
+        });
+    }
+};
+
+folly::Future<std::vector<SliceAndKey>> write_slices(
+        const std::shared_ptr<InputTensorFrame> &frame,
+        std::vector<FrameSlice>&& slices,
+        const SlicingPolicy &slicing,
+        IndexPartialKey&& key,
+        const std::shared_ptr<stream::StreamSink>& sink,
+        const std::shared_ptr<DeDupMap>& de_dup_map,
+        bool sparsify_floats) {
+    ARCTICDB_SAMPLE(WriteSlices, 0)
 
-    auto fut_writes = sink->batch_write(std::move(key_segs), de_dup_map);
+    std::vector<std::pair<FrameSlice, size_t>> slice_and_rowcount;
+    slice_and_rowcount.reserve(slices.size());
+    size_t slice_num_for_column = 0;
+    std::optional<size_t> first_row;
+    for(const auto& slice : slices) {
+        if (!first_row)
+            first_row = slice.row_range.first;
 
-    ARCTICDB_SUBSAMPLE_DEFAULT(WriteSlicesWait)
-    return std::move(fut_writes).thenValue([slices = std::move(slices)](auto &&keys) mutable {
-        std::vector<SliceAndKey> res;
-        res.reserve(keys.size());
-        for (std::size_t i = 0; i < res.capacity(); ++i) {
-            res.emplace_back(SliceAndKey{slices[i], std::move(to_atom(keys[i]))});
-        }
-        return res;
-    });
+        if (slice.row_range.first == first_row.value())
+            slice_num_for_column = 0;
+
+        slice_and_rowcount.emplace_back(slice, slice_num_for_column);
+        ++slice_num_for_column;
+    }
```

Comment:
I'm confused by this logic. `first_row` is always set to `slices[0].row_range.first` and then never modified. So `slice_num_for_column` will be zero for all slices in the first row-slice, then 1, 2, 3,... for subsequent slices?

Reply:
It's quite niche. If the data is not contiguous and we need to call the FlattenHelper we multiply the slice_num by the regular_slice_size by the array strides, to work out where our starting point is. So if a uint64 array was backwards and had a stride of -8, and we were dividing it into slices of two rows each, the first slice would start at 0 (the origin pointer) and work backwards, and then next slice would start at -16.

I think in earlier times this logic was a bit closer together, it's not obvious where it goes these days. Not totally sure how to fix it though, we definitely have the information here, and we need it to go there.